library/alloc/src/boxed.rs

```diff
@@ -62,6 +62,11 @@
 //! T` obtained from [`Box::<T>::into_raw`] may be deallocated using the
 //! [`Global`] allocator with [`Layout::for_value(&*value)`].
 //!
+//! For zero-sized values, the `Box` pointer still has to be [valid] for reads and writes and
+//! sufficiently aligned. In particular, casting any aligned non-zero integer to a raw pointer
+//! produces a valid pointer, but a pointer pointing into previously allocated memory that since got
+//! freed is not valid.
```

Comment:
```suggestion
//! produces a valid pointer, [unless it is a pointer pointing into previously allocated memory that since got
//! freed][`::core::ptr`].
```

Otherwise it may not be 100% clear which of the two rules "win"s in case of, for instance:

```rust
dealloc(ptr : *mut ()); // deallocated
Box::<()>::from_raw(ptr as usize as *mut ())
//                  ^^^^^^^^^^^^
//                  aligned non-zero integer
```

Reply:
I think I should probably change the text to say "integer literal", would that also solve your concern?